search nicknames (issue 2240)

daquexian opened the pull request:
PR for issue #2240 

According to the [document](https://developer.android.com/guide/topics/providers/contacts-provider.html), nicknames and email addresses are stored separately in different rows. So I query the `CONTACT_ID` at first and query email address according to `CONTACT_ID`

philipwhiuk commented on k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientLoader.java:
```diff
@@ -177,9 +177,46 @@ private void fillContactDataFromQuery(String query, List<Recipient> recipients,
+         * Look for contacts whose nickname contains given string, retrieve CONTACT_ID and nickname
+         * of them
+         */
+        Cursor cursorForNickname = contentResolver.query(queryUriForNickname,
```
If you pulled this into a well-named method you could get rid of the comment.

philipwhiuk commented on k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientLoader.java:
```diff
@@ -177,9 +177,46 @@ private void fillContactDataFromQuery(String query, List<Recipient> recipients,
+        /**
+         * query email etc. according to CONTACT_ID, and specify the name of recipient
+         */
+        if (cursorForNickname != null && cursorForNickname.moveToFirst()) {
```
And you could do create a method for this logic too.

Smaller methods are easier to read, the logic is more obvious and so less prone to error.

philipwhiuk commented on k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientLoader.java:
```diff
@@ -195,13 +232,24 @@ private void fillContactDataFromQuery(String query, List<Recipient> recipients,
+    }
+
+    private void fillContactDataFromCursor(Cursor cursor, List<Recipient> recipients,
+                                           Map<String, Recipient> recipientMap, String name) {
```
Maybe it's worth marking String as `@Nullable`

philipwhiuk commented on k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientLoader.java:
```diff
@@ -177,9 +177,46 @@ private void fillContactDataFromQuery(String query, List<Recipient> recipients,
+                PROJECTION_NICKNAME,
+                ContactsContract.CommonDataKinds.Nickname.NAME + " LIKE ? AND " +
+                        Data.MIMETYPE + " = '" +
+                        ContactsContract.CommonDataKinds.Nickname.CONTENT_ITEM_TYPE + "'",
```
Pass `ContactsContract.CommonDataKinds.Nickname.CONTENT_ITEM_TYPE` in as a parameter - defeats the point of parameterised queries if you don't do it where you can.

daquexian commented:
@philipwhiuk Done :) I created a method `getNicknameCursor`, but there is an error under strict mode that the cursor isn't released. But I do release it in another method `fillContactDataFromNickname`, should I ignore it?

philipwhiuk commented on k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientLoader.java:
```diff
@@ -173,13 +174,58 @@ private static String getContactIdFromContactUri(Uri contactUri) {
+                nicknameCursor.moveToNext();
+            }
+
+            nicknameCursor.close();
```
You aren't calling close() if the cursor is empty.

daquexian commented:
@philipwhiuk Sorry I forgot it.. I have fixed it and squashed the commits into one :)

Valodim commented on k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientLoader.java:
```diff
@@ -173,13 +174,62 @@ private static String getContactIdFromContactUri(Uri contactUri) {
+    private Cursor getNicknameCursor(String nickname) {
+        Uri queryUriForNickname = ContactsContract.Data.CONTENT_URI;
+
+        final String[] PROJECTION_NICKNAME =
```
should be a static field

Valodim commented on k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientLoader.java:
```diff
@@ -173,13 +174,62 @@ private static String getContactIdFromContactUri(Uri contactUri) {
+            return;
+        }
+
+        if (nicknameCursor.moveToFirst()) {
```
This can simply be `while (nicknameCursor.moveToNext())`

Valodim commented on k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientLoader.java:
```diff
@@ -173,13 +174,62 @@ private static String getContactIdFromContactUri(Uri contactUri) {
+            }
+        }
+
+        nicknameCursor.close();
```
should be in try.. finally block

Valodim commented on k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientLoader.java:
```diff
@@ -173,13 +174,62 @@ private static String getContactIdFromContactUri(Uri contactUri) {
+        Uri queryUri = ContactsContract.CommonDataKinds.Email.CONTENT_URI;
+        final ContentResolver contentResolver = getContext().getContentResolver();
+
+        final int INDEX_CONTACT_ID = 0;
```
static fields

Valodim commented on k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientLoader.java:
```diff
@@ -173,13 +174,62 @@ private static String getContactIdFromContactUri(Uri contactUri) {
+    }
+
+
     private void fillContactDataFromQuery(String query, List<Recipient> recipients,
```
This method confused me because it 1) delegates part of the filling to `fillContactDataFromNickname`, and 2) does part of the filling itself with a query. The second part should be extracted into its own method, which should leave `fillContactDataFromQuery` only calling to these two methods, which gives a clearer picture of what it does.

daquexian commented:
@Valodim Thanks for your review! I have updated it and squashed the commits :)

Besides extract a method, a another change is that for [this method](https://github.com/k9mail/k-9/pull/2264/files#diff-fb5ec3b5fcd0004c2924df6fb5581cd6R209), I make the [nickname method](https://github.com/k9mail/k-9/pull/2264/files#diff-fb5ec3b5fcd0004c2924df6fb5581cd6R217) and [another method](https://github.com/k9mail/k-9/pull/2264/files#diff-fb5ec3b5fcd0004c2924df6fb5581cd6R252) return a boolean value indicating whether there is a contact matched

Valodim commented:
Sorry for the delay. from a quick glance this looks good now. I'll need to test and look at it again, will try to do that very soon. Thanks for your work!

Valodim commented:
I just tested this and it didn't actually work for me. I created a contact, added a nickname, and then searched both for a prefix or an infix of the nickname, but the result of the contacts provider query was empty. Am I doing something wrong maybe?

daquexian commented:
@Valodim Could you please try it again? :)

Valodim commented on k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientLoader.java:
```diff
@@ -173,35 +183,107 @@ private static String getContactIdFromContactUri(Uri contactUri) {
             Map<String, Recipient> recipientMap) {
 
-        ContentResolver contentResolver = getContext().getContentResolver();
+        boolean getDataFromNickname = fillContactDataFromNickname(query, recipients, recipientMap);
```
I'd suggest
```java
boolean foundValidCursor = false;
foundValidCursor |= fillContactDataFromNickname();
foundValidCursor |= fillContactDataFromNameAndEmail();
```

daquexian commented:
@Valodim , made the required changes and squashed the commits :) In my opinion putting the contacts queried by nickname on the top is OK (For me I also want them to be on the top :) )

Valodim commented:
thanks! :+1:
